tools/src/main/scala/scala/scalanative/codegen/CodeGen.scala

```diff
@@ -5,70 +5,72 @@ import java.{lang => jl}
 import java.nio.ByteBuffer
 import java.nio.file.Paths
 import scala.collection.mutable
-import scalanative.util.{ShowBuilder, unsupported}
+import scalanative.util.{Scope, ShowBuilder, unsupported}
 import scalanative.io.{VirtualDirectory, withScratchBuffer}
 import scalanative.optimizer.analysis.ControlFlow.{Graph => CFG, Block, Edge}
 import scalanative.nir._
 
 object CodeGen {
 
   /** Generate code for given assembly. */
-  def apply(config: tools.Config, assembly: Seq[Defn]): Unit = {
-    val env = assembly.map(defn => defn.name -> defn).toMap
-
-    def debug(): Unit = {
-      val batches = mutable.Map.empty[String, mutable.Buffer[Defn]]
-      assembly.foreach { defn =>
-        val top = defn.name.top.id
-        val key =
-          if (top.startsWith("__")) top
-          else if (top == "main") "__main"
-          else {
-            val pkg = top.split("\\.").init.mkString(".")
-            if (pkg == "") "__empty"
-            else pkg
+  def apply(config: tools.Config, assembly: Seq[Defn]): Unit =
+    Scope { implicit in =>
+      val env     = assembly.map(defn => defn.name -> defn).toMap
+      val workdir = VirtualDirectory.real(config.workdir)
+
+      def debug(): Unit = {
+        val batches = mutable.Map.empty[String, mutable.Buffer[Defn]]
+        assembly.foreach { defn =>
+          val top = defn.name.top.id
+          val key =
+            if (top.startsWith("__")) top
+            else if (top == "main") "__main"
+            else {
+              val pkg = top.split("\\.").init.mkString(".")
+              if (pkg == "") "__empty"
+              else pkg
+            }
+          if (!batches.contains(key)) {
+            batches(key) = mutable.UnrolledBuffer.empty[Defn]
           }
-        if (!batches.contains(key)) {
-          batches(key) = mutable.UnrolledBuffer.empty[Defn]
+          batches(key) += defn
+        }
+        batches.par.foreach {
+          case (k, defns) =>
+            val impl =
+              new Impl(config.target, env, defns, workdir)
+            val outpath = k + ".ll"
+            withScratchBuffer { buffer =>
+              impl.gen(buffer)
+              buffer.flip
+              workdir.write(Paths.get(outpath), buffer)
+            }
         }
-        batches(key) += defn
       }
-      batches.par.foreach {
-        case (k, defns) =>
-          val impl =
-            new Impl(config.target, env, defns, config.targetDirectory)
-          val outpath = k + ".ll"
+
+      def release(): Unit = {
+        withScratchBuffer { buffer =>
+          val defns   = assembly
+          val impl    = new Impl(config.target, env, defns, workdir)
+          val outpath = "out.ll"
           withScratchBuffer { buffer =>
```

Comment:
Why the double nested scratch buffer? The outer one doesn't touch the `buffer` at all